turbo/stages/headerdownload/header_algos.go

```diff
@@ -183,122 +183,93 @@ func (hd *HeaderDownload) markPreverified(link *Link) {
-			link := hd.addHeaderAsLink(segment[i], false /* pesisted */)
-			if prevLink == nil {
-				newAnchor.links = append(newAnchor.links, link)
-			} else {
-				prevLink.next = append(prevLink.next, link)
-			}
-			prevLink = link
-			if _, ok := hd.preverifiedHashes[link.hash]; ok {
-				hd.markPreverified(link)
-			}
+		if newAnchor.blockHeight > 0 {
+			hd.anchors[newAnchorHeader.ParentHash] = newAnchor
+			heap.Push(hd.anchorQueue, newAnchor)
+		}
+	}
+	hd.removeAnchor(anchor)
+	// Add all headers in the segments as links to this anchor
+	var prevLink *Link
+	for i := len(segment) - 1; i >= 0; i-- {
+		link := hd.addHeaderAsLink(segment[i], false /* pesisted */)
+		if prevLink == nil {
+			newAnchor.links = append(newAnchor.links, link)
+		} else {
+			prevLink.next = append(prevLink.next, link)
 		}
-		prevLink.next = anchor.links
-		anchor.links = nil
-		if anchorPreverified {
-			// Mark the entire segment as preverified
-			hd.markPreverified(prevLink)
+		prevLink = link
+		if _, ok := hd.preverifiedHashes[link.hash]; ok {
+			hd.markPreverified(link)
 		}
-		return !preExisting, nil
 	}
-	return false, fmt.Errorf("extendDown attachment anchors not found for %x", anchorHash)
+	prevLink.next = anchor.links
+	anchor.links = nil
+	if anchorPreverified {
+		// Mark the entire segment as preverified
+		hd.markPreverified(prevLink)
+	}
+	return !preExisting
 }
 
 // Connect connects some working trees using anchors of some, and a link of another
-func (hd *HeaderDownload) connect(segment ChainSegment) ([]PenaltyItem, error) {
-	// Find attachment link again
-	linkH := segment[len(segment)-1]
-	// Find attachement anchors again
-	anchorHash := segment[0].Hash
-	attachmentLink, ok1 := hd.getLink(linkH.Header.ParentHash)
-	if !ok1 {
-		return nil, fmt.Errorf("connect attachment link not found for %x", linkH.Header.ParentHash)
-	}
-	if attachmentLink.preverified && len(attachmentLink.next) > 0 {
-		return nil, fmt.Errorf("cannot connect to preverified link %d with children", attachmentLink.blockHeight)
-	}
-	anchor, ok2 := hd.anchors[anchorHash]
-	if !ok2 {
-		return nil, fmt.Errorf("connect attachment anchors not found for %x", anchorHash)
-	}
+func (hd *HeaderDownload) connect(segment ChainSegment, attachmentLink *Link, anchor *Anchor) []PenaltyItem {
 	anchorPreverified := false
 	for _, link := range anchor.links {
 		if link.preverified {
 			anchorPreverified = true
 			break
 		}
 	}
-	// Anchor is removed from the map, and from the priority queue
-	delete(hd.anchors, anchor.parentHash)
-	heap.Remove(hd.anchorQueue, anchor.idx)
+	hd.removeAnchor(anchor)
 	// Iterate over headers backwards (from parents towards children)
 	prevLink := attachmentLink
 	for i := len(segment) - 1; i >= 0; i-- {
 		link := hd.addHeaderAsLink(segment[i], false /* persisted */)
+		// If we attach to already persisted link, mark this one for insertion
+		if prevLink.persisted {
+			hd.insertList = append(hd.insertList, link)
+		}
```

Comment:
In the previous implementation, you added to the hd.insertList queue only one end of the segment (the lowest). Here you are adding all the segment headers to the queue. 
I think the previous implementation was fine because in the InsertHeaders at [this ](https://github.com/ledgerwatch/erigon/blob/62d524f6010cc1ed37214c511459a1f27029b6d7/turbo/stages/headerdownload/header_algos.go#L710) line after persisting a link you add all its children to the insertQueue

Reply:
I like the deletion of the code that searched the attachmentLink again.

About the question of the insertionList, surely the reader no longer asks himself "why only the first", but the code in the InsertHeaders alredy does

```
		if len(link.next) > 0 {
			hd.insertList = append(hd.insertList, link.next...)
		}
```

so I think each link will be processed 2 times.